internal/pkg/service/appsproxy/proxy/apphandler/authproxy/selector.go

```diff
@@ -0,0 +1,183 @@
+
+		// Get path for redirect after sign in, it must not refer to an external URL
+		query := make(url.Values)
+		callback := req.URL.Query().Get(callbackQueryParam)
+		if callback != "" && strings.HasPrefix(callback, "/") {
+			query.Set(callbackQueryParam, callback)
+		}
+
+		// Render sign in page, set callback after login
+		s.redirect(w, req, selected.SignInPath(), query)
+		return nil
+	}
+
+	// Render the page, if there is no cookie or the value is invalid
+	s.pageWriter.WriteSelectorPage(w, req, status, s.selectorPageData(req, handlers))
+	return nil
+}
+
+func (s *Selector) selectorPageData(req *http.Request, handlers map[provider.ID]*Handler) *pagewriter.SelectorPageData {
+	// Pass link back to the current page, if reasonable, otherwise the user will be redirected to /
+	var callback string
+	if req.Method == http.MethodGet {
+		callback = req.URL.Path
+	}
+
+	// Base URL for all providers
+	pageURL := s.url(req, selectionPagePath, nil)
+
+	// Generate link for each providers
+	data := &pagewriter.SelectorPageData{}
+	for _, handler := range handlers {
+		query := make(url.Values)
+		query.Set(providerQueryParam, handler.ID().String())
+		if callback != "" {
+			query.Set(callbackQueryParam, callback)
+		}
+		data.Providers = append(data.Providers, pagewriter.ProviderData{
+			Name: handler.Name(),
+			URL:  pageURL.ResolveReference(&url.URL{RawQuery: query.Encode()}).String(),
+		})
+	}
+
+	// Sort items
+	slices.SortStableFunc(data.Providers, func(a, b pagewriter.ProviderData) int {
+		return strings.Compare(a.Name, b.Name)
+	})
+
+	return data
+}
+
+func (s *Selector) redirect(w http.ResponseWriter, req *http.Request, path string, query url.Values) {
+	w.Header().Set("Location", s.url(req, path, query).String())
+	w.WriteHeader(http.StatusFound)
+}
+
+func (s *Selector) url(req *http.Request, path string, query url.Values) *url.URL {
+	return &url.URL{Scheme: s.config.API.PublicURL.Scheme, Host: requtil.HostPort(req), Path: path, RawQuery: query.Encode()}
+}
+
+func (s *Selector) providerIDFromCookie(req *http.Request) provider.ID {
+	if cookie, _ := req.Cookie(providerCookie); cookie != nil && cookie.Value != "" {
+		return provider.ID(cookie.Value)
+	}
+	return ""
+}
+
+func (s *Selector) clearCookie(w http.ResponseWriter, req *http.Request) {
+	http.SetCookie(w, s.cookie(req, "", -1))
+}
+
+func (s *Selector) setCookie(w http.ResponseWriter, req *http.Request, handler *Handler) {
+	http.SetCookie(w, s.cookie(req, handler.ID().String(), handler.CookieExpiration()))
+}
+
+func (s *Selector) cookie(req *http.Request, value string, expires time.Duration) *http.Cookie {
+	v := &http.Cookie{
+		Name:     providerCookie,
+		Value:    value,
+		Path:     "/",
+		Domain:   requtil.Host(req),
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: http.SameSiteLaxMode,
+	}
+
+	if expires > 0 {
+		// If there is an expiration, set it
+		v.Expires = s.clock.Now().Add(expires)
+	} else {
+		// Otherwise clear the cookie
+		v.MaxAge = -1
+	}
```

Comment:
As I wrote (maybe you missed it in other comments)

> It better match the logic, ... if there is an expiration ... set it ... else .... clear the cookie. I will add comment.

-------------

I really don't understand your request. For me, it is important that the code is 1. functional 2. easy to understand. What I think this code accomplishes.

I have the feeling that you prefer rules over purpose: `if possible, always set the default value, don't use "else", and never otherwise`.

In my opinion, you also need to think about a person, what will be work on it after me/you, and he should notice this important part of the the function.

:arrow_right: so please let's discuss it outside this PR, please add link to the comment to the [your document](https://keboola.atlassian.net/wiki/spaces/ENGG/pages/3301507091/TMP+Martin+Va+ko+Onboarding+Suggestions).